msrestazure/azure_active_directory.py

```diff
@@ -657,29 +658,108 @@ class MSIAuthentication(BasicTokenAuthentication):
     - resource (str): Alternative authentication resource, default
       is 'https://management.core.windows.net/'.
 
-    :param int port: MSI local port if VM/VMSS context (ignored otherwise)
-
     .. versionadded:: 0.4.14
     """
 
     def __init__(self, port=50342, **kwargs):
         super(MSIAuthentication, self).__init__(None)
 
+        if port != 50342:
+            warnings.warn("The 'port' argument is no longer used, and will be removed in a future release", DeprecationWarning)
         self.port = port
+
         self.msi_conf = {k:v for k,v in kwargs.items() if k in ["client_id", "object_id", "msi_res_id"]}
 
         self.cloud_environment = kwargs.get('cloud_environment', AZURE_PUBLIC_CLOUD)
         self.resource = kwargs.get('resource', self.cloud_environment.endpoints.active_directory_resource_id)
 
-    def set_token(self):
         if _is_app_service():
             if self.msi_conf:
                 raise AuthenticationError("User Assigned Entity is not available on WebApp yet.")
+        elif "MSI_ENDPOINT" not in os.environ:
+            # Use IMDS if no MSI_ENDPOINT
+            self._vm_msi = _ImdsTokenProvider(self.resource, self.msi_conf)
+
+    def set_token(self):
+        if _is_app_service():
             self.scheme, _, self.token = get_msi_token_webapp(self.resource)
-        else:
+        elif "MSI_ENDPOINT" in os.environ:
```

Comment:
Will it be possible that `MSI_ENDPOINT` could be an empty string?

Reply:
Seems a broken env for me, we should keep it simple here

Reply:
sure